sympy/solvers/ode/ode.py

```diff
@@ -1752,6 +1752,115 @@ def match_2nd_linear_bessel(r, func):
     rn['b4'] = point
     return rn
 
+
+# find coefficients of terms f(t), diff(f(t),t) and higher derivatives
+# and similarly for other functions g(t), diff(g(t),t) in all equations.
+# Here j denotes the equation number, funcs[l] denotes the function about
+# which we are talking about and k denotes the order of function funcs[l]
+# whose coefficient we are calculating.
+def linearity_check(eqs, j, func, funcs, func_coef, order, t, is_linear_):
+    for k in range(order[func] + 1):
+        func_coef[j, func, k] = collect(eqs.expand(), [diff(func, t, k)]).coeff(diff(func, t, k))
```

Comment:
This function modifies `func_coef` which is passed as an input argument. It is generally confusing to modify the input arguments like this. I think it would be better to have a function for linearity checking that finds and returns the resulting coefficients.

Perhaps the function could be something like:
```
def get_linear_coeffs(eq, funcs):
    # eq is an equation e.g. x(t).diff(t) + 2*x(t) + 3*y(t) + t
    # funcs is a list of the dependent variables e.g. [x(t), y(t)]
    ...
    if not linear:
        return None
    else:
        return coeffs # {x(t).diff(t): 1, x(t):2, y(t): 3, 1: t}
```